### PR TITLE
Show samples grouped by framework

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
We have a lot of samples and `$ appbuilder sample` lists all of them. It is difficult to find the one you need. Add separate commands to improve this behavior:
* `appbuilder sample hybrid` - lists only Cordova samples
* `appbuilder sample native` - lists only NativeScript samples
* `appbuilder sample website` - lists only Mobile Website samples
* `appbuilder sample` - lists all samples, grouped by Framework. For each framwork the samples are listed grouped by categories.

Update common lib, so added two methods to PrompterStubs implementation. Add help for new commands.

### CODE REVIEW

**Comment by rosen-vladimirov** [comment]
I'll consider using cli-table for one of our next releases, for the moment I prefer to keep current formatting.

File: lib/services/samples-service.ts [hunk 1]
```diff
@@ -20,13 +20,16 @@ class Sample {
...
-	private static GITHUB_TELERIK_LOCATION_ENDPOINT = "https://api.github.com/orgs/telerik/repos?per_page=100";
 	private static GITHUB_REGEX = /https:\/\/github[.]com\/Icenium\/(?!deprecated-)(sample-|.*?-sample-)[\w\W]+[.]git$/i;
 	private static NAME_FORMAT_REGEX = /(sample-|-)/gi;
 	private static NAME_PREFIX_REMOVAL_REGEX = /(sample-)/i;
 	private static REMOTE_LOCK_STATE_PRIVATE = "private";
 	private static SAMPLES_PULL_FAILED_MESSAGE = "Failed to retrieve samples list. Please try again a little bit later.";
 	private static NATIVESCRIPT_SAMPLE_CUTENESS_NAME = "nativescript-sample-cuteness";
+	private static GITHUB_CORDOVA_SAMPLES_REGEX = new RegExp("https:\/\/github[.]com\/Icenium\/(?!deprecated-)sample-[\\w\\W]+[.]git$", "i");
+	private static GITHUB_MOBILE_WEBSITE_SAMPLES_REGEX = new RegExp("https:\/\/github[.]com\/Icenium\/(?!deprecated-)mobilewebsite-sample-[\\w\\W]+[.]git$", "i");
+	private static GITHUB_NS_SAMPLES_REGEX = new RegExp("https:\/\/github[.]com\/Icenium\/(?!deprecated-)nativescript-sample-[\\w\\W]+[.]git$", "i");
```
Comment (rosen-vladimirov) [inline lib/services/samples-service.ts]: remove (?!deprecated-) part from this Regex and from all related except GITHUB_REGEX. samples starting with deprecated can never match this regex.

File: lib/services/samples-service.ts [hunk 2]
```diff
@@ -20,13 +20,16 @@ class Sample {
...
+	//private static GITHUB_ALL_SAMPLES_REGEX = new RegExp("https:\/\/github[.]com\/Icenium\/sample-[\\w\\W]+[.]git$|https:\/\/github[.]com");
```
Comment (rosen-vladimirov) [inline lib/services/samples-service.ts]: remove commented code

File: lib/commands/samples.ts [hunk 3]
```diff
@@ -49,3 +49,39 @@ class CloneCommandParameter implements ICommandParameter {
 		}).future<boolean>()();
 	}
 }
+
+class PrintCordovaSamplesInformationCommand implements ICommand {
```
Comment (ligaz) [inline lib/commands/samples.ts]: Is it possible to create a generic command for this and eliminate the duplicated code with just passing the framework identifier and having 3 registrations for each framework?

File: lib/services/samples-service.ts [hunk 4]
```diff
@@ -38,23 +41,26 @@ export class SamplesService implements ISamplesService {
 		private $errors: IErrors,
 		private $fs: IFileSystem,
 		private $httpClient: Server.IHttpClient,
-		private $staticConfig: IStaticConfig) {
+		private $staticConfig: IStaticConfig,
+		private $projectConstants: Project.IProjectConstants) {
 	}
 
-	private get samples(): IFuture<Sample[]> {
+	public printSamplesInformation(platform?: string): IFuture<void> {
```
Comment (ligaz) [inline lib/services/samples-service.ts]: Is it a `platform` or a `framework`? :smiley_cat:

File: lib/commands/samples.ts (hunk 3 above)
Comment (rosen-vladimirov) [inline lib/commands/samples.ts]: I'll try to remove most of the duplicate code, but we prefer to have the commands separated (to have separate command for each framework) as it makes it easier to show correct help and information about the framework (context sensitive help).

File: lib/services/samples-service.ts [hunk 5]
```diff
@@ -122,28 +128,41 @@ export class SamplesService implements ISamplesService {
...
 
-	private getAllSamples(): IFuture<Sample[]> {
+	private getSamples(platform?: string): IFuture<Sample[]> {
 		return (() => {
-			var iceniumOrganizationSamples = this.getRepositories(SamplesService.GITHUB_ICENIUM_LOCATION_ENDPOINT, (repo: any) => SamplesService.GITHUB_REGEX.test(repo.clone_url) && !repo[SamplesService.REMOTE_LOCK_STATE_PRIVATE]).wait();
-			var telerikOrganizationSamples = this.getRepositories(SamplesService.GITHUB_TELERIK_LOCATION_ENDPOINT, (repo: any) => repo.name === SamplesService.NATIVESCRIPT_SAMPLE_CUTENESS_NAME).wait();
-			var repos = _.union(iceniumOrganizationSamples, telerikOrganizationSamples);
+			var regex: RegExp;
+			platform = platform || "";
+			switch(platform.toLowerCase()) {
```
Comment (ligaz) [inline lib/services/samples-service.ts]: You can extract the switch to a separate method. Will make the code more readable :eyeglasses:

Comment (rosen-vladimirov) [inline lib/services/samples-service.ts]: already done this, you have to refresh GitHub page :)

File: lib/commands/samples.ts (hunk 3 above)
Comment (rosen-vladimirov) [inline lib/commands/samples.ts]: I've tried using BaseCommand for all of them and use constructor injection for projectConstants and samplesService in it, but in PrintCordovaSamplesInformationCommand I'll have to inject both of them in order to call super(...):
```TypeScript
class PrintCordovaSamplesInformationCommand extends PrintBaseCommand {
	constructor(protected $samplesService: ISamplesService,
		protected $projectConstants: Project.IProjectConstants) {
		super($samplesService, $projectConstants);
	}
```
So we still have much of the code duplicated. In this case I prefer to keep the current code.

**Comment by teobugslayer** [comment]
:+1: after green build

**Comment by rosen-vladimirov** [comment]
ping @ikoevska for the help content

File: lib/services/samples-service.ts [hunk 6]
```diff
@@ -122,28 +126,46 @@ export class SamplesService implements ISamplesService {
...
-		return (() => {
-			var iceniumOrganizationSamples = this.getRepositories(SamplesService.GITHUB_ICENIUM_LOCATION_ENDPOINT, (repo: any) => SamplesService.GITHUB_REGEX.test(repo.clone_url) && !repo[SamplesService.REMOTE_LOCK_STATE_PRIVATE]).wait();
-			var telerikOrganizationSamples = this.getRepositories(SamplesService.GITHUB_TELERIK_LOCATION_ENDPOINT, (repo: any) => repo.name === SamplesService.NATIVESCRIPT_SAMPLE_CUTENESS_NAME).wait();
-			var repos = _.union(iceniumOrganizationSamples, telerikOrganizationSamples);
+	private getRegExpForFramework(framework?: string): RegExp {
+		var regex: RegExp;
+		framework = framework || "";
+		switch(framework.toLowerCase()) {
+			case this.$projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript.toLowerCase():
+				regex = SamplesService.GITHUB_NS_SAMPLES_REGEX;
```
Comment (ligaz) [inline lib/services/samples-service.ts]: You can just return here. That's was one of my internal reasons for the method extraction :smile:

Comment (rosen-vladimirov) [inline lib/services/samples-service.ts]: :facepalm: (I really need this emoticon for this case) 
I'll fix it, thanks for noticing this.

File: lib/commands/samples.ts [hunk 7]
```diff
@@ -4,26 +4,36 @@
...
+var subCommands = {
+	"*list": "",
+	"hybrid": frameworkIdentifiers.Cordova,
+	"native": frameworkIdentifiers.NativeScript,
+	"website": frameworkIdentifiers.MobileWebsite
+};
+var samplesService = $injector.resolve("samplesService");
+_.each(subCommands, (subCommand, framework) => {
+	$injector.registerCommand("sample|" + subCommand, new PrintSamplesCommand(samplesService, framework));
+});
```
Comment (Fatme) [inline lib/commands/samples.ts]: I suggest you to use something similar to this service https://github.com/Icenium/icenium-cli/blob/master/lib/services/project-commands-service.ts#L8. I also think it should be great if you create `CommandsServiceBase` and reuse the logic.

**Comment by teobugslayer** [comment]
:+1:

File: docs/man_pages/project/creation/sample-hybrid.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+sample hybrid
+==========
+
+Usage | Syntax
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-hybrid.md]: Usage | Synopsis

File: docs/man_pages/project/creation/sample-hybrid.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+------|-------
+General | `$ appbuilder sample hybrid`    
+
+Lists all available Cordova sample apps with name, description, GitHub repository, and clone command. To clone a selected sample app,
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-hybrid.md]: Lists all available Apache Cordova sample apps with name, description, GitHub repository and clone command. To clone a selected sample app, run its clone command as it is listed.

File: docs/man_pages/project/creation/sample-hybrid.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+[create hybrid](create-hybrid.html) | Creates a new project from an Apache Cordova-based template.
+[create native](create-native.html) | Creates a new project from a NativeScript-based template.
+[create website](create-website.html) | Creates a new project from a Mobile Website-based template.
+[create](create.html) | Creates a project for hybrid or native development.
+[init hybrid](init-hybrid.html) | Initializes an existing Apache Cordova project for development in the current directory.
+[init native](init-native.html) | Initializes an existing NativeScript project for development in the current directory.
+[init website](init-website.html) | Initializes an existing Mobile Website project for development in the current directory.
+[init](init.html) | Initializes an existing project for development.
+[sample](sample.html) | Lists all available sample apps.
+[sample native] (sample-native.html) | Lists all available NativeScript sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-hybrid.md]: Extra space between ] and (

File: docs/man_pages/project/creation/sample-hybrid.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+[sample website] (sample-website.html) | Lists all available Mobile Website sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-hybrid.md]: Extra space between ] and (

File: docs/man_pages/project/creation/sample-native.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
+sample native
+==========
+
+Usage | Syntax
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-native.md]: Usage | Synopsis

File: docs/man_pages/project/creation/sample-native.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+------|-------
+General | `$ appbuilder sample native`    
+
+Lists all available NativeScript sample apps with name, description, GitHub repository, and clone command. To clone a selected sample app,
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-native.md]: Lists all available NativeScript sample apps with name, description, GitHub repository and clone command. To clone a selected sample app, run its clone command.

File: docs/man_pages/project/creation/sample-native.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+[create hybrid](create-hybrid.html) | Creates a new project from an Apache Cordova-based template.
+[create native](create-native.html) | Creates a new project from a NativeScript-based template.
+[create website](create-website.html) | Creates a new project from a Mobile Website-based template.
+[create](create.html) | Creates a project for hybrid or native development.
+[init hybrid](init-hybrid.html) | Initializes an existing Apache Cordova project for development in the current directory.
+[init native](init-native.html) | Initializes an existing NativeScript project for development in the current directory.
+[init website](init-website.html) | Initializes an existing Mobile Website project for development in the current directory.
+[init](init.html) | Initializes an existing project for development.
+[sample](sample.html) | Lists all available sample apps.
+[sample hybrid] (sample-hybrid.html) | Lists all available Cordova sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-native.md]: Again the extra spaces. And please, change all "available Cordova sample apps" to "available Apache Cordova sample apps"

File: docs/man_pages/project/creation/sample-website.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
+sample website
+==========
+
+Usage | Syntax
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-website.md]: Usage | Synopsis

File: docs/man_pages/project/creation/sample-website.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+------|-------
+General | `$ appbuilder sample website`    
+
+Lists all available Mobile Website sample apps with name, description, GitHub repository, and clone command. To clone a selected sample app,
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-website.md]: Lists all available mobile website sample apps with name, description, GitHub repository and clone command. To clone a selected sample app, run its clone command.

File: docs/man_pages/project/creation/sample-website.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+[create hybrid](create-hybrid.html) | Creates a new project from an Apache Cordova-based template.
+[create native](create-native.html) | Creates a new project from a NativeScript-based template.
+[create website](create-website.html) | Creates a new project from a Mobile Website-based template.
+[create](create.html) | Creates a project for hybrid or native development.
+[init hybrid](init-hybrid.html) | Initializes an existing Apache Cordova project for development in the current directory.
+[init native](init-native.html) | Initializes an existing NativeScript project for development in the current directory.
+[init website](init-website.html) | Initializes an existing Mobile Website project for development in the current directory.
+[init](init.html) | Initializes an existing project for development.
+[sample](sample.html) | Lists all available sample apps.
+[sample native] (sample-native.html) | Lists all available NativeScript sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample-website.md]: Empty spaces =)

File: docs/man_pages/project/creation/sample.md [hunk 18]
```diff
@@ -3,13 +3,18 @@ sample
 
 Usage | Synopsis
 ------|-------
-General | `$ appbuilder sample [<Command>]`
+General | `$ appbuilder sample`
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: Leave the original synopsis here

File: docs/man_pages/project/creation/sample.md [hunk 19]
```diff
@@ -3,13 +3,18 @@ sample
...
+Additional | `$ appbuilder sample [<Command>]`
 
-Lists all available sample apps with name, description, GitHub repository and clone command.
+Lists all available sample apps with name, description, GitHub repository, and clone command. To clone a selected sample app,
+run its clone command as listed by `$ appbuilder sample`
 
 ### Attributes
-`<Command>` extends the `sample` command. You can set the following values for this attribute.
-* `clone` - Clones the selected sample app.
+`<Command>` is any related command that extends the sample command. You can run the following related commands:
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: \`\<Command\>\` extends the \`sample\` command. You can set the following values for this attribute.

File: docs/man_pages/project/creation/sample.md [hunk 20]
```diff
@@ -3,13 +3,18 @@ sample
...
+* `hybrid` - Lists all available Cordova sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: Apache Cordova

File: docs/man_pages/project/creation/sample.md [hunk 21]
```diff
@@ -27,4 +32,9 @@ Command | Description
 [init native](init-native.html) | Initializes an existing NativeScript project for development in the current directory.
 [init website](init-website.html) | Initializes an existing Mobile Website project for development in the current directory.
 [sample](sample.html) | Lists all available sample apps with name, description, GitHub repository, and clone command.
+[sample clone](sample-clone.html) | Clones the selected sample app from GitHub to your local file system.
+[sample native] (sample-native.html) | Lists all available NativeScript sample apps.
+[sample hybrid] (sample-hybrid.html) | Lists all available Cordova sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: Apache Cordova

File: docs/man_pages/project/creation/sample-hybrid.md (hunk 10 above)
Comment (rosen-vladimirov) [inline docs/man_pages/project/creation/sample-hybrid.md]: wow, you're good :smile_cat:

File: docs/man_pages/project/creation/sample.md (hunk 18 above)
Comment (rosen-vladimirov) [inline docs/man_pages/project/creation/sample.md]: Done :)

**Comment by rosen-vladimirov** [comment]
run ci

**Comment by ligaz** [comment]
:+1: The refactoring came out pretty nice.

File: docs/man_pages/project/creation/sample.md [hunk 22]
```diff
@@ -3,13 +3,18 @@ sample
 
 Usage | Synopsis
 ------|-------
-General | `$ appbuilder sample [<Command>]`
+General | `$ appbuilder sample`    
+Additional | `$ appbuilder sample [<Command>]`
 
-Lists all available sample apps with name, description, GitHub repository and clone command.
+Lists all available sample apps with name, description, GitHub repository, and clone command. To clone a selected sample app,
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: Lists all available sample apps with name, description, GitHub repository and clone command. To clone a selected sample app, run its clone command as listed by `$ appbuilder sample`

File: docs/man_pages/project/creation/sample.md [hunk 23]
```diff
@@ -3,13 +3,18 @@ sample
...
+run its clone command as listed by `$ appbuilder sample`
 
 ### Attributes
 `<Command>` extends the `sample` command. You can set the following values for this attribute.
-* `clone` - Clones the selected sample app.
+* `hybrid` - Lists all available Apache Cordova sample apps.
+* `native` - Lists all available NativeScript sample apps.
+* `website` - Lists all available Mobile Website sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: mobile website

File: docs/man_pages/project/creation/sample.md [hunk 24]
```diff
@@ -27,4 +32,8 @@ Command | Description
 [init native](init-native.html) | Initializes an existing NativeScript project for development in the current directory.
 [init website](init-website.html) | Initializes an existing Mobile Website project for development in the current directory.
 [sample](sample.html) | Lists all available sample apps with name, description, GitHub repository, and clone command.
+[sample clone](sample-clone.html) | Clones the selected sample app from GitHub to your local file system.
+[sample native](sample-native.html) | Lists all available NativeScript sample apps.
+[sample hybrid](sample-hybrid.html) | Lists all available Apache Cordova sample apps.
+[sample website](sample-website.html) | Lists all available Mobile Website sample apps.
```
Comment (ikoevska) [inline docs/man_pages/project/creation/sample.md]: mobile website

**Comment by ikoevska** [comment]
:+1:

**Pull request merged**